Add createSubspaceProvider, mirroring Redux's createProvider.

conartist6 opened the pull request:
| Q                       | A 
| ----------------------- | ---
| Fixed Issues?           | Fixes #89
| Minor: New Feature?     |  👍
| Tests Added + Pass?     | Yes

See changed documentation for a description of the changes : )

mpeyper commented:
Nice work on this.  I've left a few comments for discussion, but overall happy with this approach.

As a side note for your intended use case, [the way react-redux uses context is changing soon](https://github.com/reduxjs/react-redux/issues/950) so it might not continue to work going forward, but it's too soon to tell on any of that, so just a heads up.

conartist6 commented on packages/react-redux-subspace/src/components/SubspaceProvider.js:
```diff
@@ -10,37 +10,41 @@ import React, { Children } from 'react'
 import { subspace }  from 'redux-subspace'
 
-class SubspaceProvider extends React.PureComponent {
+export function createSubspaceProvider(storeKey = 'store', parentStoreKey = 'store') {
```
Arrow function, sure. I considered doing that actually. My opinion is that if you want that behavior the argument order should be swapped:
``` 
export const createSubspaceProvider = (parentStoreKey = 'store', storeKey = parentStoreKey, ) => {
```
Actually though any way you slice it someone is going to have to specify a default parameter they don't really care about in order to specify one that they do. Maybe a literate syntax like  `createProvider().withStoreKey(key)` would be best of all?

mpeyper replied:
I think I agree about the defaulting order.  If the parent store key is changed, you probably want the store key to be changed too.

The scenarios I see:

1. parent store default, child store default
   ```js
   const SubspaceProvider = createSubspaceProvider()
   ```
2. parent store override, child store override
   ```js
   const SubspaceProvider = createSubspaceProvider('other')
   ```
3. parent store default, child store override
   ```js
   const SubspaceProvider = createSubspaceProvider('store', 'other')
   ```
4. parent store override, child store default
   ```js
   const SubspaceProvider = createSubspaceProvider('other', 'store')
   ```
5. parent store override, child store different override
   ```js
   const SubspaceProvider = createSubspaceProvider('other1', 'other2')
   ```

As you said, in both 3. and 4. one has be be defined as the default key, which kinda sucks.

How about
```js
export const createSubspaceProvider = ({ storeKey = 'store', parentStoreKey = storeKey, childStoreKey = storeKey } = {}) => {
```

1. parent store default, child store default
   ```js
   const SubspaceProvider = createSubspaceProvider()
   ```
2. parent store override, child store override
   ```js
   const SubspaceProvider = createSubspaceProvider({ storeKey: 'other' })
   ```
3. parent store default, child store override
   ```js
   const SubspaceProvider = createSubspaceProvider({ childStoreKey: 'other' })
   ```
4. parent store override, child store default
   ```js
   const SubspaceProvider = createSubspaceProvider({ parentStoreKey: 'other' })
   ```
5. parent store override, child store different override
   ```js
   const SubspaceProvider = createSubspaceProvider({ parentStoreKey: 'other1', childStoreKey: 'other2' })
   ```

With this, you'd only specify the one you want to override, with `storeKey` allowing both to be overridden. 

Or, possibly, accepting either a string or an object.  If a string is provider, it sets both `parentStoreKey` and `childStoreKey` to be the string, something like:

```js
const DEFAULT_STORE_KEY = 'store'

const createSubspaceProvider = (storeKey = DEFAULT_STORE_KEY) => {
  const { parentStoreKey = DEFAULT_STORE_KEY, childStoreKey = DEFAULT_STORE_KEY } = typeof(storeKey) === 'string' 
    ? { parentStoreKey: storeKey, childStoreKey: storeKey } 
    : storeKey

  // ...
}

1. parent store default, child store default
   ```js
   const SubspaceProvider = createSubspaceProvider()
   ```
2. parent store override, child store override
   ```js
   const SubspaceProvider = createSubspaceProvider('other')
   ```
3. parent store default, child store override
   ```js
   const SubspaceProvider = createSubspaceProvider({ childStoreKey: 'other' })
   ```
4. parent store override, child store default
   ```js
   const SubspaceProvider = createSubspaceProvider({ parentStoreKey: 'other' })
   ```
5. parent store override, child store different override
   ```js
   const SubspaceProvider = createSubspaceProvider({ parentStoreKey: 'other1', childStoreKey: 'other2' })
   ```

conartist6 replied:
Yeah I think your last suggestion hits the nail on the head. It's verbose to do the thing that it should be difficult/rare to need to do, and when you're just trying to proxy on a setting that you already configured using Redux `createProvider` the syntax is actually exactly the same as createProvider's.

Perfect.

conartist6 commented on packages/react-redux-subspace/src/components/SubspaceProvider.js:
```diff
@@ -10,37 +10,41 @@ import React, { Children } from 'react'
-    store: PropTypes.object.isRequired
-}
+    SubspaceProvider.childContextTypes = {
+        [storeKey]: PropTypes.object
```
Haha ok I'll find out.

conartist6 commented:
Yeah I've read though the threads on context. I think this should be even easier to do once we're just passing around {Provider, Consumer} pairs.

conartist6 commented:
OK, so the `[contextKey]Provider` context key is used for ReactRedux Subscriptions which help ensure that parents components have their updates handled before children. I have updated the API and incorporated that context key into the changeset.

mpeyper commented on packages/react-redux-subspace/src/components/SubspaceProvider.js:
```diff
@@ -10,37 +10,50 @@ import React, { Children } from 'react'
-    getChildContext() {
-        const makeSubspaceDecorator = (props) => props.subspaceDecorator || subspace(props.mapState, props.namespace)
+export function createSubspaceProvider(paramStoreKey = 'store') {
+    let storeKey = typeof paramStoreKey === 'object' ? paramStoreKey.storeKey || 'store' : paramStoreKey;
```
I think I would much prefer to be more explicit about what each `storeKey` is when supplying an object to prevent any confusion.

```js
const DEFAULT_STORE_KEY = 'store'

export const createSubspaceProvider = (storeKey = DEFAULT_STORE_KEY) => {
  const childStoreKey = typeof storeKey === 'object' ? storeKey.childStoreKey || DEFAULT_STORE_KEY : storeKey
  const parentStoreKey = typeof storeKey === 'object' ? storeKey.parentStoreKey || DEFAULT_STORE_KEY : storeKey;

  class SubspaceProvider extends React.PureComponent {
    getChildContext() {
      const makeSubspaceDecorator = (props) => props.subspaceDecorator || subspace(props.mapState, props.namespace)

      return {
        [childStoreKey]: makeSubspaceDecorator(this.props)(this.context[parentStoreKey]),
        [`${childStoreKey}Subscription`]: this.context[`${parentStoreKey}Subscription`],
      }
    }

    render() {
      return Children.only(this.props.children)
    }
}
```

mpeyper commented on packages/react-redux-subspace/src/components/SubspaceProvider.js:
```diff
@@ -10,37 +10,50 @@ import React, { Children } from 'react'
+
+            if (storeKey !== 'store') {
+                console.log('parentStoreKey: ', parentStoreKey)
+                console.log(this.context[parentStoreKey])
```
Did you mean to include these?

conartist6 replied:
oops, nope...

mpeyper commented on packages/react-redux-subspace/src/components/SubspaceProvider.js:
```diff
@@ -10,37 +10,50 @@ import React, { Children } from 'react'
+            }
+            return {
+                [storeKey]: makeSubspaceDecorator(this.props)(this.context[parentStoreKey]),
+                [`${storeKey}Subscription`]: this.context[`${parentStoreKey}Subscription`],
```
I'm starting to wonder if it would be easier to just use `react-redux`'s `Provider` to inject our subspaced store back into the context instead of trying to replicate how `Provider` does it.  It might make us more resilient to any changes they may make in the future.  

I won't insist on changing it to do that as part of this PR as I don't anticipate them changing anytime soon (other than the huge context API changes but we'll likely have to overhaul a bunch of stuff when that happens anyway).

mpeyper commented on packages/react-redux-subspace/test/components/SubspaceProvider-spec.js:
```diff
@@ -37,6 +37,74 @@ describe('SubspaceProvider Tests', () => {
+
+        expect(testComponent.text()).to.equal("expected")
+    })
         
```
wouldn't mind also seeing some cases for the defaulting of the keys when only one is provided, i.e.

* `createSubspaceProvider({ childStoreKey })`
* `createSubspaceProvider({ parentStoreKey })`

mpeyper commented:
I'm going to close this due to inactivity and the react-redux v6 changes proposed in #103 will make essentially make changing the store key in the legacy context obsolete.
